File: app/src/main/java/uk/co/jatra/lifecycledialog/LifeCycleDialogFragment.java
package uk.co.jatra.lifecycledialog;

import android.app.Dialog;
import android.arch.lifecycle.ViewModelProviders;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;

import static uk.co.jatra.lifecycledialog.AnswerModel.Answer.NO;
import static uk.co.jatra.lifecycledialog.AnswerModel.Answer.NO_ANSWER;
import static uk.co.jatra.lifecycledialog.AnswerModel.Answer.YES;

public class LifeCycleDialogFragment extends DialogFragment {

    private AnswerModel answerModel;
    private AnswerModel.Answer answer = NO_ANSWER;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        answerModel = ViewModelProviders.of(getActivity()).get(AnswerModel.class);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        android.app.AlertDialog.Builder builder = new android.app.AlertDialog.Builder(getContext());
        builder.setMessage("Have you rotated recently?");
        builder.setPositiveButton("Yes!", (dialogInterface, i) -> {
            answer = YES;
        });
        builder.setNegativeButton("No", (dialogInterface, i) -> {
            answer = NO;
        });
        return builder.create();
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);
        answerModel.answer(answer);
    }
}

File: app/src/main/java/uk/co/jatra/lifecycledialog/LifeCycleDialogFragment.kt
package uk.co.jatra.lifecycledialog

import android.app.Dialog
import android.arch.lifecycle.ViewModelProviders
import android.content.DialogInterface
import android.os.Bundle
import android.support.v4.app.DialogFragment
import uk.co.jatra.lifecycledialog.Answer.*

class LifeCycleDialogFragment : DialogFragment() {

    private var answerModel: AnswerModel? = null
    private var answer: Answer = NO_ANSWER

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        answerModel = ViewModelProviders.of(activity).get(AnswerModel::class.java)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = android.app.AlertDialog.Builder(context)
        builder.setMessage("Have you rotated recently?")
        builder.setPositiveButton("Yes!") { dialogInterface, i -> answer = YES }
        builder.setNegativeButton("No") { dialogInterface, i -> answer = NO }
        return builder.create()
    }

    override fun onDismiss(dialog: DialogInterface?) {
        super.onDismiss(dialog)
        answerModel!!.answer(answer)
    }
}